overlord/install/install.go

```diff
@@ -193,3 +206,232 @@ func CheckEncryptionSupport(model *asserts.Model, tpmMode secboot.TPMProvisionMo
 
 	return res.Type, res.UnavailableErr
 }
+
+// BuildInstallObserver creates an observer for gadget assets if
+// applicable, otherwise the returned gadget.ContentObserver is nil.
+// The observer if any is also returned as non-nil trustedObserver if
+// encryption is in use.
+func BuildInstallObserver(model *asserts.Model, gadgetDir string, useEncryption bool) (
+	observer gadget.ContentObserver, trustedObserver *boot.TrustedAssetsInstallObserver, err error) {
+
+	// observer will be a nil interface by default
+	trustedObserver, err = boot.TrustedAssetsInstallObserverForModel(model, gadgetDir, useEncryption)
+	if err != nil && err != boot.ErrObserverNotApplicable {
+		return nil, nil, fmt.Errorf("cannot setup asset install observer: %v", err)
+	}
+	if err == nil {
+		observer = trustedObserver
+		if !useEncryption {
+			// there will be no key sealing, so past the
+			// installation pass no other methods need to be called
+			trustedObserver = nil
+		}
+	}
+
+	return observer, trustedObserver, nil
+}
+
+// PrepareEncryptedSystemData executes preparations related to encrypted system data:
+// * provides trustedInstallObserver with the chosen keys
+// * uses trustedInstallObserver to track any trusted assets in ubuntu-seed
+// * save keys and markers for ubuntu-data being able to safely open ubuntu-save
+func PrepareEncryptedSystemData(model *asserts.Model, keyForRole map[string]keys.EncryptionKey, trustedInstallObserver *boot.TrustedAssetsInstallObserver) error {
+	// validity check
+	if len(keyForRole) == 0 || keyForRole[gadget.SystemData] == nil || keyForRole[gadget.SystemSave] == nil {
+		return fmt.Errorf("internal error: system encryption keys are unset")
+	}
+	dataEncryptionKey := keyForRole[gadget.SystemData]
+	saveEncryptionKey := keyForRole[gadget.SystemSave]
+
+	// make note of the encryption keys
+	trustedInstallObserver.ChosenEncryptionKeys(dataEncryptionKey, saveEncryptionKey)
+
+	// XXX is the asset cache problematic from initramfs?
```

Comment:
Thanks for this comment, I wonder if we should have something like `// TODO:<tag>` to make these questions easier to find after we done the refactor but I'm probably overthinking it.